0101-nix formatting.md

```diff
@@ -0,0 +1,65 @@
+---
+feature: nix_formatting
+start-date: 2021-08-17
+author: Raphael Megzari
+co-authors: (find a buddy later to help out with the RFC)
+shepherd-team: (names, to be nominated and accepted by RFC steering committee)
+shepherd-leader: (name to be appointed by RFC steering committee)
+related-issues: (will contain links to implementation PRs)
+---
+
+# Summary
+
+[summary]: #summary
+
+Decide on a recommended automated formatter for nix files in nixpkgs.
+
+# Motivation
+
+[motivation]: #motivation
+
+Prevent debate around how things should be formatted.
+
+# Detailed design
+
+[design]: #detailed-design
+
+The implementation of this RFC should include several parts
+
+- Agree on an automated formatter. [nixpkgs-fmt](https://github.com/nix-community/nixpkgs-fmt) is proposed here.
+- If adopted, agree on a schedule to format nixpkgs codebase. Should it be done at once, or on a per package basis.
+- Potentially agree on a hook to enforce formatting
+
+# Examples and Interactions
+
+[examples-and-interactions]: #examples-and-interactions
+
+This section is not needed for this RFC.
+
+# Drawbacks
+
+[drawbacks]: #drawbacks
+
+- Having a commit that changes the formatting, can make git blame harder to use. It will need `--ignore-rev` flag.
+- Every formatter will have bugs. An example of a bug for nixpkgs-fmt can be found [here](https://github.com/NixOS/nixpkgs/pull/129392)
+
+# Alternatives
+
+[alternatives]: #alternatives
+
+- Alternative formatter [nixfmt](https://github.com/serokell/nixfmt)
+- Keep the status quo of not having an official formatter. The danger is that this creates discord within the nix community. On top of fragmented the community it can generate lengthy discussions (which do not advance the eco-system).
+
+# Unresolved questions
+
+[unresolved]: #unresolved-questions
+
+- Not sure how much work there is left on nixpkgs-fmt before most people would consider it ok to use. Not even sure how much it is actually used.
+- Are there situation where automated formatting is worse than manual formatting ? Do we want to have exceptions to automated formatting ?
```

Comment:
I have to admit that updating code generation by putting the generator call inside a script and calling `nixpkgs-fmt` afterwards sounds like a nonzero but manageable amount of work.

On the other hand, maintaining generators to always be formatter-clean would create unncessary pushback to fixing some weird corner cases.

Reply:
IMO just fixup the generated files after they have been generated. Either the tool itself can do that, your pre-commit hooks, or the update script. Either of these works. If we enforce formatting of files then almost no file should be excluded. Especially not generated files as being able to read them is important. There is no excuse to treat those a black boxes that shouldn't be looked at.

Reply:
The original tool should be fixed to always comply, to avoid churn. Or we marke the file "no format". Or we do https://github.com/NixOS/rfcs/pull/109/